### PR TITLE
Cirque trackpad features: circular scroll, inertial cursor

### PR DESCRIPTION
**Pull request by dkao**
Adding gestures such as circular scroll and inertial cursor, along with some general QoL changes

## Description

- Circular scroll gesture - Disabled by default. Preloaded configuration detects touches originating in outer ring, after validation to differentiate between swipe from edge vs. scroll gesture, the trackpad turns into a scroll knob until finger lift.
- Inertial cursor - Disabled by default. If enabled, cursor will continue moving after a flick gesture and slows down by kinetic friction. Similar to trackball mode on Steam Controller.
- Add function to disable trackpad tap gesture.
- Recalibrate baseline after setting ADC attenuation.
- Set CPI according to trackpad diameter.
- Control cursor smoothing.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by dkao** [comment]
_pointing_device_drivers.c_ has gotten quite large, any suggestions for where the extra processing code should live? Should I move `cirque_pinnacle_get_report()` and other functions into _cirque_pinnacle.c_?
Also should I add compile flags to disable features and save memory, or would that add too much complexity to user configuration?

File: drivers/sensors/cirque_pinnacle.c [hunk 1]
```diff
@@ -138,9 +140,10 @@ void ERA_ReadBytes(uint16_t address, uint8_t* data, uint16_t count) {
         RAP_Write(ERA_CONTROL, 0x05); // Signal ERA-read (auto-increment) to Pinnacle
 
         // Wait for status register 0x1E to clear
+        timeout_timer = timer_read();
         do {
             RAP_ReadBytes(ERA_CONTROL, &ERAControlValue, 1);
-        } while (ERAControlValue != 0x00);
+        } while ((ERAControlValue != 0x00) && (timer_elapsed(timeout_timer) <= CIRQUE_PINNACLE_TIMEOUT));
```
Comment (Kriechi) [inline drivers/sensors/cirque_pinnacle.c:146]: At least on I2C we could use ` && touchpad_init==true` directly, as this will be set to false if there is any I2C error (or timeout). Not sure if SPI considers `CIRQUE_PINNACLE_TIMEOUT` at all?

File: drivers/sensors/cirque_pinnacle.c [hunk 2]
```diff
@@ -241,6 +293,8 @@ void cirque_pinnacle_init(void) {
     RAP_Write(Z_IDLE_COUNT, 5);
 
     cirque_pinnacle_set_adc_attenuation(CIRQUE_PINNACLE_ATTENUATION);
+    // Perform manual calibration after setting ADC attenuation
```
Comment (Kriechi) [inline drivers/sensors/cirque_pinnacle.c:296]: "manual calibration" sounds like the user needs to tweak a potentiometer or something. This is in fact more an "auto-calibration" or "self-calibration" routine, right?

File: drivers/sensors/cirque_pinnacle.c [hunk 3]
```diff
@@ -192,6 +197,53 @@ void cirque_pinnacle_tune_edge_sensitivity(void) {
...
+    } while ((calconfig & 0x01) && (timer_elapsed(timeout_timer) <= 200));
+
+    cirque_pinnacle_clear_flags();
+}
+
+// Enable/disable cursor smoothing, smoothing is enabled by default
+void cirque_pinnacle_cursor_smoothing(bool enable) {
+    uint8_t feedconfig3;
+
+    // FeedConfig3 (Advanced feature flags)
```
Comment (Kriechi) [inline drivers/sensors/cirque_pinnacle.c:229]: Is this register available in all trackpad versions (TM040040/TM035035/TM023023) and Pinnacle firmwares?
GT-AN-090620 does not list it at all (presumably hidden under `RESERVED`.

File: drivers/sensors/cirque_pinnacle.c (hunk 2 above)
Comment (dkao) [reply drivers/sensors/cirque_pinnacle.c:296]: Fair enough, how about rewording it as "Force a calibration after setting ADC attenuation"? Borrows terminology from GT-AN-090624 section 1.2. I would call it "force compensation" but the bit itself is called "calibrate".

Comment (Kriechi) [reply drivers/sensors/cirque_pinnacle.c:296]: _perform/trigger/force_ are all 👍

File: drivers/sensors/cirque_pinnacle.c (hunk 3 above)
Comment (dkao) [reply drivers/sensors/cirque_pinnacle.c:229]: FeedConfig3 register is listed in GT-AN-090625, which is stated as a companion document to GT-AN-090620. Same thing is stated for the ERA document (GT-AN-090623) and compensation document (GT-AN-090624).
Additionally, it is documented in Cirque's circle trackpad example code [here](https://github.com/cirque-corp/Cirque_Pinnacle_1CA027/blob/master/Additional_Examples/AnyMeas_Example/Pinnacle.h). I don't see any indication it would not be present for a particular version of circle trackpad.
Empirically, I've tested this register on TM035035 and TM040040; the effect is subtle on 40mm but more prominent on 35mm.

File: drivers/sensors/cirque_pinnacle.c (hunk 1 above)
Comment (dkao) [reply drivers/sensors/cirque_pinnacle.c:146]: SPI doesn't have ack or check for any timeout. It will return whatever state MISO happens to be floating in, if the device is not connected. That could be either 0xFF or 0x00 (or something inbetween if there's noise on the line?).
Looking at AVR's _spi_master.c_ for example: `spi_start()` always returns true, except if chip select pin is invalid but that's an error before ever interacting with the device. So `touch_init` check can't really apply for SPI.

File: drivers/sensors/cirque_pinnacle.c [hunk 4]
```diff
@@ -192,6 +197,53 @@ void cirque_pinnacle_tune_edge_sensitivity(void) {
...
+    // CalConfig1 (Compensation)
+    // Bit 0: Calibrate, 1=calibrate, 0=complete
+    // Bit 1: Background Comp Enable, 1=enable, 0=disable
+    // Bit 2: NERD Comp Enable (No, I don't know what NERD means), 1=enable, 0=disable
+    // Bit 3: Track Error Comp Enable, 1=enable, 0=disable
+    // Bit 4: Tap Comp Enable, 1=enable, 0=disable
+    // Bit 6: Calibration Matrix Disable, 1=disabled, 0=enabled
+    RAP_ReadBytes(CALIBRATION_CONFIG_1, &calconfig, 1);
+    calconfig |= 0x01;
+    RAP_Write(CALIBRATION_CONFIG_1, calconfig);
```
Comment (KarlK90) [inline drivers/sensors/cirque_pinnacle.c:214]: Same as the feedconfig_3 comment about bitfields

**Review by KarlK90 — CHANGES_REQUESTED** [review]
Thanks for continued effort to improve the cirque track-pads. Existed to try these myself (I have some heading my way). I have to review the code in greater detail later on. On thing that I would like to see addressed is the introduction of bitfield structs for all the configuration registers that get manipulated and that are currently only described indirectly. This should help with type safety and readability of the code.

**Comment by KarlK90** [comment]
> _pointing_device_drivers.c_ has gotten quite large, any suggestions for where the extra processing code should live? Should I move `cirque_pinnacle_get_report()` and other functions into _cirque_pinnacle.c_? 

I'm not sure here. Either we should go with the current approach to not mix the actual mouse report generation with the low level driver code. In this case we should split the _pointing_device_drivers.c_ into sensor types and add them under a folder. Or we move the report generation logic into the low level drivers itself. Both might be out of scope for this PR though. Also I would like to hear @drashna opinion on this.

> Also should I add compile flags to disable features and save memory, or would that add too much complexity to user configuration?

I wouldn't add that just to save memory. If a feature introduced behavior that might be unwanted it should be made opt-out either at compile time or runtime.

File: drivers/sensors/cirque_pinnacle.c [hunk 5]
```diff
@@ -192,6 +197,53 @@ void cirque_pinnacle_tune_edge_sensitivity(void) {
...
+    // FeedConfig3 (Advanced feature flags)
+    // Bit 0: DualPoint Buttons, 1=enable, 0=disable
+    // Bit 1: Smoothing Disable, 1=disable, 0=enable
+    // Bit 2: Palm/NERD measurements Disable, 1=disable, 0=enable
+    // Bit 3: Noise Avoidance Disable, 1=disable, 0=enable
+    // Bit 4: WRAP lockout Disable
+    // Bit 5: Dynamic EMI adjust Disable
+    // Bit 6: HW EMI detection Disable
+    // Bit 7: SW EMI detection Disable
+    RAP_ReadBytes(FEEDCONFIG_3, &feedconfig3, 1);
```
[the remark this answers is not in the document]
Comment (dkao) [reply drivers/sensors/cirque_pinnacle.c:238]: I can do this, but is it really the right choice for a platform independent driver?
> The order of allocation of bit-fields within a unit (high-order to low-order or low-order to high-order) is implementation-defined. The alignment of the addressable storage unit is unspecified.

As an example of a caveat: QMK would have to only ever support little endian microcontrollers (which I believe currently is the case anyway?), otherwise each register would need to have its structure defined once for little endian and then again backwards for big endian.
I think it's safer to import [Cirque's register definition file](https://github.com/cirque-corp/Cirque_Pinnacle_1CA027/blob/master/Additional_Examples/AnyMeas_Example/Pinnacle.h) and use that instead.

[the remark this answers is not in the document]
Comment (KarlK90) [reply drivers/sensors/cirque_pinnacle.c:238]: > I think it's safer to import [Cirque's register definition file](https://github.com/cirque-corp/Cirque_Pinnacle_1CA027/blob/master/Additional_Examples/AnyMeas_Example/Pinnacle.h) and use that instead.

I agree that we should just import the already existing definitions by Cirque because they provide descriptive names (which is the main problem I have with these magic numbers).

> but is it really the right choice for a platform independent driver?

We fully control the build environment which is GCC only at this point (I have a clang patch in the pipeline but not pr'ed yet) and the GCC implementation seems to work for us. Also all current (avr, arm, risc-v) and future platforms (xtensa?) that I can think of are little endian - so that would be fine.

[the remark this answers is not in the document]
Comment (dkao) [reply drivers/sensors/cirque_pinnacle.c]: I've uploaded an update with a modified version of Cirque's register definition file.
As a start I kept the wordy "HOSTREG__" prefixes to differentiate between regular registers and extended registers, let me know what you think.

File: quantum/pointing_device_drivers.c [hunk 6]
```diff
@@ -130,45 +402,57 @@ report_mouse_t cirque_pinnacle_get_report(report_mouse_t mouse_report) {
...
+
+    if (features.tap_enable) {
+        mouse_report = trackpad_tap(mouse_report, touchData);
     }
 
     return mouse_report;
 }
 
+uint16_t cirque_pinnacle_get_cpi(void) {
+    return (uint16_t)roundf((float)cirque_pinnacle_get_scale() * 25.4f / (float)CIRQUE_PINNACLE_DIAMETER_MM);
```
Comment (drashna) [inline quantum/pointing_device_drivers.c:444]: I'm not a huge fan of using floats here, since AVR is still being used. 

If it helps, the lib8tion library is present already (used for rgb stuff mostly), and that may help.

**Review by drashna — COMMENTED** [review]
Yeah, I'd prefer that the drivers file here be kept simple as possible.  

Ideally, "fancy" features (kind of like this), be handled by `pointing_device.c`, so that any device can benefit from that code. 

I do understand that the cirque is very much unique, especially in comparison to the other sensors.  I'm not sure what the best solution is, especially long term.

However, it may be best to pull out the features that would work with other sensors (the inertial cursor code, perhaps), and move that to the `pointing_devices.c` file, while device specific features (such as the circular scroll) be moved into the cirque files. 

Long term, if we get other trackpads supported, it may be worth having those features in the core (rather than sensor code), and having them behind defines.  Eg, "optical sensor" vs "trackpads" (and probably added by default).

File: quantum/pointing_device_drivers.c (hunk 6 above)
Comment (dkao) [reply quantum/pointing_device_drivers.c:444]: > However, it may be best to pull out the features that would work with other sensors (the inertial cursor code, perhaps), and move that to the pointing_devices.c file, while device specific features (such as the circular scroll) be moved into the cirque files.

The report structure currently doesn't support pulling these features to higher level code. Inertial cursor depends on knowing when a finger lifts, and circular scroll depends on absolute coordinates, both of which are not exposed by `report_mouse_t`. It might be possible to differentiate finger lift vs. stationary finger by the last velocity before `mouse_report.x` and `mouse_report.y` become 0, but that sounds pretty janky to me.
Seems like I should move these lower level to somewhere in _drivers/sensors/_ then, at least in the short term? I'm not sure about modifying the `get_report()` API for this one use case.

> I'm not a huge fan of using floats here, since AVR is still being used.
> If it helps, the lib8tion library is present already (used for rgb stuff mostly), and that may help.

For simple stuff like converting between inch and millimeter, sure I can get rid of the float usage.
I had tried earlier on to use lib8tion for the fancy trigonometry further in, but wasn't able to get it to behave correctly. Some inputs were too wide for lib8tion and had wrapped around before I could evaluate the accuracy.
Been running this as is on AVR; while it has trouble reporting at 1KHz on master side, it's still above 500Hz (tested by making inertial cursor reports run at 1ms interval and turning off friction coefficient).

**Comment by dkao** [comment]
Extracted circular scroll and tap to _drivers/sensors/cirque_pinnacle_gestures.[ch]_.
Extracted inertial cursor to _quantum/pointing_device_gestures.[ch]_, these are still called from _pointing_device_drivers.c_ but are usable by other drivers if desired.
Floating point math converted to integer/fixed point. Inertial scroll fixed point is slightly slower than floating point on AVR, should be much faster on architectures with native 32-bit words.

**Review by KarlK90 — CHANGES_REQUESTED** [review]
@dkao My cirque trackpads finally arrived and I have tested the additions! Great work and I'm looking forward to see more keyboards using these in the future.

Now to the review part:

> I wouldn't add that just to save memory. If a feature introduced behavior that might be unwanted it should be made opt-out either at compile time or runtime.

I gave this statement before actually looking at the numbers and the code, so sorry for revoking that decision so late in the process. This is due to comparing the compile size before and after this PR on AVR Pro Micros gave a difference of ~2.3KB in data, which is unfortunately the lowest denominator we have to address and huge in the world of these mcus. Therefore my proposal:

1. The gestures are opt-in at compile time.
2. The gestures are enabled at runtime by default.
3. The gestures enabled state can be switched at runtime (already implemented).

Otherwise LGTM!

File: drivers/sensors/cirque_pinnacle.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (KarlK90) [reply drivers/sensors/cirque_pinnacle.c]: Thanks! This looks good now.

File: drivers/sensors/cirque_pinnacle.c (hunk 1 above)
Comment (KarlK90) [reply drivers/sensors/cirque_pinnacle.c:114]: This actually fixed a dead lock for me that happened with the code before. I started a small refactoring of the `RAP_Read/Write` functions. If you find it useful feel free to include it in this pr [cirque_timeout.patch.txt](https://github.com/qmk/qmk_firmware/files/9073188/cirque_timeout.patch.txt)

Comment (KarlK90) [reply drivers/sensors/cirque_pinnacle.c:114]: As a side note, if a single I2C transfer fails the whole driver will turn into an invalid state - which requires a reboot of the keyboard. I can think of mechanism on the pointing device driver level that will try to re-init a driver in such a case after a certain timeout. But this is not in the scope of this PR.

File: drivers/sensors/cirque_pinnacle_gestures.c [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            return 32768;
+    }
+
+    int32_t abs_y = dy > 0 ? dy : -dy;
+    int16_t a;
+
+    if (dx >= 0)
+        a = 8192 - (8192 * (dx - abs_y) / (dx + abs_y));
+    else
+        a = 24576 - (8192 * (dx + abs_y) / (abs_y - dx));
```
Comment (Kriechi) [inline drivers/sensors/cirque_pinnacle_gestures.c:73]: how about curly braces here to prevent a future "goto fail"?
or does the QMK coding style not mention anything about this?

File: drivers/sensors/cirque_pinnacle_gestures.c [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+#else
+                scroll.axis = x < 0;
+#endif
+            }
+        } else if (scroll.state == SCROLL_DETECTING) {
+            report.suppress_touch = true;
+            // already detecting scroll, check movement from touchdown location
+            mag = sqrt16((x - scroll.x) * (x - scroll.x) + (y - scroll.y) * (y - scroll.y));
+            if (mag >= scroll.config.trigger_px) {
+#define ABS(a) (a > 0 ? a : -a)
```
Comment (Kriechi) [inline drivers/sensors/cirque_pinnacle_gestures.c:128]: any tradeoff to consider when moving this to a static-inline function? mixing both "styles" with defines and inline function feels a bit inconsistent (although ABS is very short... so maybe ok?)

also: is there no ABS in any of the available math libs?

Comment (dkao) [reply drivers/sensors/cirque_pinnacle_gestures.c]: Thanks for bringing this up! Apparently I'd spend too long in an environment where abs() is overloaded with the version for doubles, and completely forgot the stdlib integer version even existed.

File: drivers/sensors/cirque_pinnacle.c (hunk 1 above)
Comment (dkao) [reply drivers/sensors/cirque_pinnacle.c:114]: I think handling `touch_init` and I2C transaction failures adds too much to this PR, let's handle that in a separate one.

**Comment by dkao** [comment]
Added enable flags for these three gestures, to be defined in `config.h`. Tap gesture changes from default enabled to opt-in.

**Review by drashna — APPROVED** [review]

**Review by KarlK90 — APPROVED** [review]
@dkao Thanks for addressing my remarks so fast! Any final changes before the merge?

**Comment by dkao** [comment]
@KarlK90 Thanks, I have some changes queued up to handle POINTING_DEVICE_COMBINED for circular scroll. Was thinking to open up a separate PR since this one has already been reviewed. Can add them here if you don't mind.
Stats:
```
 drivers/sensors/cirque_pinnacle_gestures.c | 29 +++++++++++++++++------------
 quantum/pointing_device.c                  |  2 +-
 2 files changed, 18 insertions(+), 13 deletions(-)
```
